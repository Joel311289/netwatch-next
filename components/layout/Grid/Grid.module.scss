/** Grid (main component) **/

.wrapper {
  box-sizing: border-box;
}

/** GridItem (child component) **/

.item {
  width: 100%;
}
