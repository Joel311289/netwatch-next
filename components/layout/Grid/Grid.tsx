import { createContext, useContext, useMemo } from 'react';
import { ResponsiveStyleValue } from '@mui/system';
import { GridSize, GridSpacing } from '@mui/material/Grid/Grid';
import GridMUI from '@mui/material/Grid';
import { ElementChildren, ElementHTML } from '@/types';
import { withChildrenFiltered } from '@/hoc/withChildrenFiltered';
import { classes } from '@/utils/helpers';
import styles from '@/components/layout/Grid/Grid.module.scss';

/* -------------------------------------------------------------------------- */
/** Context **/
/* -------------------------------------------------------------------------- */

type GridContextProps = typeof defaultValue;

const defaultValue = {};

const GridContext = createContext<GridContextProps>(defaultValue);

const useGridContext = () => useContext(GridContext);

/* -------------------------------------------------------------------------- */
/** GridItem (child component) **/
/* -------------------------------------------------------------------------- */

type GridItemProps = ElementHTML &
  ElementChildren<JSX.Element> & {
    xs?: boolean | GridSize;
    sm?: boolean | GridSize;
    md?: boolean | GridSize;
    lg?: boolean | GridSize;
    xl?: boolean | GridSize;
  };

const GridItem = ({ children, className, xs, sm, md, lg, xl }: GridItemProps) => {
  useGridContext();

  return (
    <GridMUI
      item
      className={classes(styles.item, className)}
      xs={xs}
      sm={sm}
      md={md}
      lg={lg}
      xl={xl}>
      {children}
    </GridMUI>
  );
};

/* -------------------------------------------------------------------------- */
/** Grid (main component) **/
/* -------------------------------------------------------------------------- */

export type GridProps = typeof defaultProps &
  ElementHTML &
  ElementChildren<Array<JSX.Element>> & {
    itemsPerRow?: number;
    spacing?: ResponsiveStyleValue<GridSpacing> | Array<ResponsiveStyleValue<GridSpacing>>;
  };

const defaultProps = {
  children: [],
  itemsPerRow: 1,
  spacing: 0.5
};

const Grid = ({ children, className, spacing }: GridProps) => {
  const rowSpacing = useMemo(() => (Array.isArray(spacing) ? spacing[0] : spacing), [spacing]);
  const columnSpacing = useMemo(() => (Array.isArray(spacing) ? spacing[1] : spacing), [spacing]);

  return (
    <GridContext.Provider value={{}}>
      <GridMUI
        role="grid"
        className={classes(styles.wrapper, className)}
        columns={60}
        container
        rowSpacing={rowSpacing as ResponsiveStyleValue<GridSpacing>}
        columnSpacing={columnSpacing as ResponsiveStyleValue<GridSpacing>}>
        {children}
      </GridMUI>
    </GridContext.Provider>
  );
};

Grid.defaultProps = defaultProps;

const GridWithChildrenFiltered = withChildrenFiltered(Grid, { Item: GridItem });

export default GridWithChildrenFiltered;
