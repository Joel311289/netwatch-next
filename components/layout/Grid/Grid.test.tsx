import { render, screen } from '@testing-library/react';
import Grid from '@/components/layout/Grid/Grid';

const items = [...Array(3)].map((_, index) => <Grid.Item key={index}>item ${index + 1}</Grid.Item>);

describe('Tests Grid component', () => {
  test('renders component correctly', () => {
    const { container } = render(<Grid>{items}</Grid>);

    expect(container).toMatchSnapshot();
  });

  test('renders children correctly', () => {
    const gap = 5;
    render(<Grid gap={gap}>{items}</Grid>);

    expect(screen.getByRole('grid').childNodes.length).toBe(items.length);
  });

  test('renders any children correctly', () => {
    const spacing = [5, 10];
    render(<Grid spacing={spacing} />);

    expect(screen.getByRole('grid').childNodes.length).toBe(0);
  });
});
