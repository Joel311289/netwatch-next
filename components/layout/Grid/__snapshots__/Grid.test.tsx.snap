// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Grid component renders component correctly 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container wrapper css-125mmfg-MuiGrid-root"
    role="grid"
  >
    <div
      class="MuiGrid-root MuiGrid-item item css-13i4rnv-MuiGrid-root"
    >
      item $
      1
    </div>
    <div
      class="MuiGrid-root MuiGrid-item item css-13i4rnv-MuiGrid-root"
    >
      item $
      2
    </div>
    <div
      class="MuiGrid-root MuiGrid-item item css-13i4rnv-MuiGrid-root"
    >
      item $
      3
    </div>
  </div>
</div>
`;
